### ISSUE TITLE
[Enhancement] Consider removing --disable-web-security and migrate to Qt 5.15.2

### ISSUE THREAD

**Issue by shlemiel** [dvorka/mindforger#1506]
To make WebEngine to be able to access local files, additional parameters are passed to the QApplication. However, In Qt 5.15.2, passing the above parameters to the QApplication does not seem like working.

It seems like that for a webpage to be able to read local files (e.g. images), without disabling web security, the uri scheme should be same with the file path. To quickly resolve this, changing QUrl parameter to "file://" could be considered.

Briefly tested for macOS Monterey 12.6.3 / Windows 10 with Qt 5.15.2.
+ Also tested for macOS Monterey 12.6.3 / Windows 10 with Qt 5.9.9.

https://github.com/dvorka/mindforger/compare/dev/1.55.1...shlemiel:mindforger_poc0:dev/1.55.1

**Comment by dvorka** [comment]
@shlemiel thank you for the suggestion 🙏 I will definitely give it a try - hopefully URL change will help!

I wanted to upgrade Qt a few months ago and spent a lot of time trying to make it work on all platforms, but I failed to make it work on macOS. I tried many combinations of various WebEngine parameters, but was not able to find a way how to load images (as the security was more strict).

I tried WebEngine Qt 5.12.0 / Chromium ???  @ macOS BigSur ... at that time.

**Comment by shlemiel** [comment]
Thank you for the concern.

In addition, there was a bug affecting IME behavior in old versions of Qt WebEngine. To summarize the case, I have to explain how the unicode IME (specifically, the Korean IME) works and what causes the problem. Unicode IMEs generally have an intermediate state that handles characters that are not completely constructed yet. In this situation, when note preview is turned on while editing a note, the periodical update of note preview steals the focus of IME, which results that intermediately constructed character just completes right after the update, remain unfinished.

However, the problem seems to be disappeared after I upgraded kit to Qt 5.15.2. I searched the issue a bit but could not found any about it. Anyway, I hope this could also fix the above issue.

**Comment by shlemiel** [comment]
For a side note, in Qt 5.15.2, I tried to change WebKit to WebEngine in Ubuntu 22.04, and felt no big performance degradation in that environment. The first loading of WebEngine seemed to be slow (~ 2 secs), but after that moment, it seemed to work quite fast. But this may be needed to be confirmed. If you want to test it against Ubuntu, below env variable need to be added. (Without the below, WebEngine does not work in Ubuntu 22.04.)

```
qputenv("QTWEBENGINE_CHROMIUM_FLAGS", "--no-sandbox") ;
```

**Comment by dvorka** [comment]
@shlemiel thank you for all the suggestions - these insights are extremely valuable! Gordic tie of settings, environment variables and configuration around WebKit/Engine is crazy...

**Comment by asmwarrior** [comment]
> Thank you for the concern.
> 
> In addition, there was a bug affecting IME behavior in old versions of Qt WebEngine. To summarize the case, I have to explain how the unicode IME (specifically, the Korean IME) works and what causes the problem. Unicode IMEs generally have an intermediate state that handles characters that are not completely constructed yet. In this situation, when note preview is turned on while editing a note, the periodical update of note preview steals the focus of IME, which results that intermediately constructed character just completes right after the update, remain unfinished.
> 
> However, the problem is resolved by upgrading kit to Qt 5.15.2. I searched the issue a bit but could not found any about it anyway.

Hi, I see similar issue here when using 1.54.0, which use QT 5.9.9. When I try to type some Chinese chars in the IME, when the preview window is open, sometimes, I have not finished typing the whole word, such as I would like to type "xing" as a Chinese word, but the "xi" will be entered. This is annoying when typing. Are there any new release version I can use? Maybe a nightly build version?

Thanks.

**Comment by asmwarrior** [comment]
I try to download a nightly build version from this link: [Wingman: adding cURL, adding Wingman dialog skeleton #1514](https://ci.appveyor.com/project/dvorka/mindforger/builds/48896524)

After installing, I see the QT is still 5.9.9. Any good news about the QT 5.15.2? Thanks.

**Comment by dvorka** [comment]
@asmwarrior thank you for trying the nightly and your interest in MindForger!

I would like to upgrade Qt, I tried in the past (but was not successful on all the platforms - new versions bring fixes, but also new :lady_beetle:). Anyway, I plan to give it a try prior 2.0.0 release:

* Which OS/distribution do you use? Windows?

I'm thinking about experimentally releasing one platform on new Qt and perhaps it would be yours.

---

My personal notes:

* Qt upgrade justification: internationalization.
* Possible solution to Qt security sandbox restrictions (images/files on new Qt versions): `qputenv("QTWEBENGINE_CHROMIUM_FLAGS", "--no-sandbox") ;`

**Comment by asmwarrior** [comment]
> @asmwarrior thank you for trying the nightly and your interest in MindForger!
> 
> I would like to upgrade Qt, I tried in the past (but was not successful on all the platforms - new versions bring fixes, but also new 🐞). Anyway, I plan to give it a try prior 2.0.0 release:
> 
>     * Which OS/distribution do you use? Windows?

Yes, I'm using Windows. Both Windows 7 64bit and Windows 10 64bit. I think the Qt 5.x support both the two OS.

> 
> 
> I'm thinking about experimentally releasing one platform on new Qt and perhaps it would be yours.
> 
> My personal notes:
> 
>     * Qt upgrade justification: internationalization.
> 
>     * Possible solution to Qt security sandbox restrictions (images/files on new Qt versions): `qputenv("QTWEBENGINE_CHROMIUM_FLAGS", "--no-sandbox") ;`

If you have a new nightly build version, I'd grad to try it. Thanks.

BTW: when I use the mindforger for the first time, I found it is hard to use it as a beginner, because I don't know how to add a new section. After watch the youtube video, I finally know how to do it. So I suggest you should put the youtube video in your github homepage, especially in the first place.

**Comment by dvorka** [comment]
> Yes, I'm using Windows...

👍 BTW, my plan is to make Windows build green first (my primary OS is Linux and MF 2.0.0 brings many changes) - including tests and successful installer build for the current `Qt 5.9.*`. Once I have a stable distro, I will start experiments with new(er) `Qt 5.*.*` versions.

> If you have a new nightly build version, I'd grad to try it.

🙏 I will let you know as soon as it's ready.

> I found it is hard to use it as a beginner, because I don't know how to add a new section...

Thank you for letting me know - it is important feedback. It seems to me that only nerds are able to use MindForger and it is difficult to understand for common users. It's obviously my fault - primarily I created it for myself and shared it with the OSS community + I'm really bad in documentation/YT videos (as I'm not enjoying it as much as programming). Anyway, I know that it is really needed and will try to do better. Thank you again!

**Comment by dvorka** [comment]
MindForger tests are in green again https://ci.appveyor.com/project/dvorka/mindforger/builds/49106012/artifacts (Qt 5.9.*) ... I can start with the experiments...

**Comment by asmwarrior** [comment]
> MindForger Windows build & tests are in green again https://ci.appveyor.com/project/dvorka/mindforger/builds/49106012/artifacts (Qt 5.9.*) ... I can start with the experiments...

With this version, I'd like to report another issue, that is the font is blue when the "Edit" is not enabled.

See this image below:

![image](https://github.com/dvorka/mindforger/assets/561818/da55ff51-d9ef-4729-8574-fe60813346a4)

When the Edit is enabled, the text is clean, see the below image. I'm not sure why.

![image](https://github.com/dvorka/mindforger/assets/561818/c122cc8b-3c77-42d2-8ff0-264ac4d89784)

**Comment by dvorka** [comment]
<img width="1440" alt="Screenshot 2024-02-04 at 16 45 55" src="https://github.com/dvorka/mindforger/assets/2090143/91052d8d-99b4-4e95-80e0-61da442087d6">

@asmwarrior you are genius! 😉 ... a few years ago, when I was trying to make new Qt versions work on macOS, I created the desperate test you can see above in the screenshot. I used your POC, disabled the code in the main class `mindforger.cpp` which was used to tweak (Qt/vars/pars) ... and it works! As you described above, I don't see significant performance degradation.

I'm experimenting with Qt 5.15.2 on macOS as I use a Windows VM and the latest 5.x.x Qt installation takes a lot of time. I expect that it will work as well.

Ad **blur**: this is really weird - hopefully it will be fixed by the Qt upgrade.

**Comment by asmwarrior** [comment]
Hi, I looked at the nightly build history here:

https://ci.appveyor.com/project/dvorka/mindforger/history

It looks like there is no nightly build for the Windows with the QT 5.1x version. I see all the red flags in the recent commits. Hopefully you will fix those build issue, so I can test them. Thanks.

**Comment by dvorka** [comment]
@asmwarrior yes, I didn't succeed to make the installer build successful on AppVeyor by the end of weekend and now I'm back in daily routine (work / family). New Qt requires new Microsoft Visual Studio which clearly needs changes in the MF installer. I will take some time as Win is not my primary Os. Please stay tuned.

**Comment by dvorka** [comment]
<img width="1920" alt="mf-2 0 0-qt5-15-2" src="https://github.com/dvorka/mindforger/assets/2090143/bdde7276-b513-4aa9-8163-52e2c11dda34">

MF 2.0.0 @ Qt 5.15.2 @ Windows 10 w/ working images.

**Comment by dvorka** [comment]
@asmwarrior this is installer local installer build - please give it a try:

* https://www.mindforger.com/TRANSFER/mindforger-2.0.0-rc-e678c24.exe

Thank you!

**Comment by asmwarrior** [comment]
> @asmwarrior this is local installer build - please give it a try:
> 
>     * https://www.mindforger.com/TRANSFER/mindforger-2.0.0-rc-e678c24.exe
> 
> 
> Thank you!

Hi, I'm grad to tell you that this version works OK under my Win7 64bit system. It does not have the font blur issue as I mentioned before, also I don't see it lose the focus when I'm using the Chinese IME input. Great job!

**Comment by dvorka** [comment]
@asmwarrior 💪 thank you for testing of the release candidate! I really appreciate
 your interest in MindForger, suggestions, ideas and time you invested in the project. Thank YOU!

**Comment by asmwarrior** [comment]
> @asmwarrior 💪 thank you for testing of the release candidate! I really appreciate your interest in MindForger, suggestions, ideas and time you invested in the project. Thank YOU!

I see one issue, when I use this MindForger to open a markdown file(created by other markdown file editors), the application hangs. This does not happen in the markdown file created by the MindForger.

**Comment by dvorka** [comment]
@asmwarrior Can you please share (somehow anonymized) example of such Markdown file so that I can reproduce it and fix the :lady_beetle: ?

**Comment by dvorka** [comment]
@asmwarrior I'm closing this issue as the original :lady_beetle: has been fixed/resolved (with your help - thank you for that!) in MindForger 2.0.0 which was just released. I created new GitHub issue for Markdown parsing bug you reported in your last message: https://github.com/dvorka/mindforger/issues/1535

Thanks again!

**Comment by asmwarrior** [comment]
> > @asmwarrior this is local installer build - please give it a try:
> > ```
> > * https://www.mindforger.com/TRANSFER/mindforger-2.0.0-rc-e678c24.exe
> > ```
 
> > Thank you!
> 
> Hi, I'm grad to tell you that this version works OK under my Win7 64bit system. It does not have the font blur issue as I mentioned before, also I don't see it lose the focus when I'm using the Chinese IME input. Great job!

I'd like to report the font blur issue again.

I found a very interesting issue:

case 1, if I use a Windows 7 64bit PC, and run the MindForger there, if I view the desktop directly on the PC's screen, the font is blur.
case 2, if I use the windows remote desktop client from a Windows 10 64bit, and access the Windows 7 which mention in the above case 1, the font is not blur.

I have tried several methods to try to solve the font blur issue in Win7, but all failed.

The method I tried are: 
1, use a resource editor tool to tweak the manifest of the exe file to enable or disable the DPI scaling feature, no success.
2, add some environment variable such as: `set QT_AUTO_SCREEN_SCALE_FACTOR=0 or 1` or `set QT_SCALE_FACTOR=0 or 1`, and later start the mindforger again, no success.

So, I give up.
